### ISSUE TITLE
GCC error output lags mintty badly

### ISSUE THREAD

**Issue by mbudnek** [mintty/mintty#712]
I recently updated mintty from 2.7.9 to 2.8.1, and ever since then GCC's error and warning output has caused mintty to scroll very slowly.  If there's a lot of errors it can sometimes cause mintty to take several seconds to redraw the screen.  I haven't encountered any text besides GCC output that causes this to happen, but it happens both when compiling something locally and when compiling things remotely using ssh.

```
$ mintty --version
mintty 2.8.1 (x86_64-pc-cygwin)
© 2013/2017 Andy Koppe / Thomas Wolff
License GPLv3+: GNU GPL version 3 or later
There is no warranty, to the extent permitted by law.
```

And my .minttyrc:
```
BoldAsFont=yes
CursorType=block
CursorBlinks=no
FontHeight=10
Columns=120
Rows=75
Term=xterm-256color
ForegroundColour=197,200,198
BackgroundColour=29,31,33
Black=7,54,66
BoldBlack=0,43,54
Red=220,50,47
BoldRed=203,75,22
Green=0,200,75
BoldGreen=0,200,75
Yellow=204,204,100
BoldYellow=204,204,120
Blue=102,153,204
BoldBlue=102,153,204
Magenta=200,100,200
BoldMagenta=200,100,200
Cyan=42,161,152
BoldCyan=42,161,152
White=238,232,213
BoldWhite=253,246,227
BellType=0
BellTaskbar=no
CopyOnSelect=no
Font=DejaVu Sans Mono for Powerline
```

**Comment by mintty** [comment]
Can you provide a screen log as a test case? I.e. a plain file to be output with cat that exhibits the problem?

**Comment by mbudnek** [comment]
Sure, here's a couple screenfulls of output from a couple different test files (I had to zip it to get github to accept the attachment): [gcc.zip](https://github.com/mintty/mintty/files/1459477/gcc.zip).  Just extract gcc.out from the zip and cat that.  Then try to scroll the screen or just hit enter a few times.

Strangely enough, the issue doesn't seem quite as bad since downgrading to 2.7.9 and upgrading back to 2.8.1.  It's still there, but manifests more as just mild slowdown than the massive slowdown I was seeing before.

**Comment by mintty** [comment]
There was one change in 2.8.0 that might have an impact in theory (the commit to fix issue #680).
Please exchange the quote marks ‘ and ’ in your test file for straight quotes ' each and check whether that makes a difference.
It's about width checking for ambiguous-width characters; I could introduce a cache for that purpose...

**Comment by mbudnek** [comment]
Yep, replacing the fancy quotes with straight quotes seems to eliminate the slowdown.

**Comment by mintty** [comment]
Two additional questions, please:
What if you test with the original contents and mintty option `-o FontRender=textout`?
Are you running on a slow machine?

**Comment by mbudnek** [comment]
`FontRender=textout` doesn't seem to have any affect.

No, the machine I'm seeing this on is pretty capable: I7-7700K CPU and 32GB of RAM.

**Comment by mintty** [comment]
I assume you mean with `FontRender=textout` you still see the lag.
If you can self-compile mintty, please try the following test patch; does it make a difference?

--- wintext.c~  2017-11-10 12:49:30.155609000 +0000
+++ wintext.c   2017-11-10 17:22:26.192380900 +0000
@@ -2082,6 +2082,7 @@ win_check_glyphs(wchar *wcs, uint num)
 int
 win_char_width(xchar c)
 {
+return 1;
   int findex = (term.curs.attr.attr & FONTFAM_MASK) >> ATTR_FONTFAM_SHIFT;
   struct fontfam * ff = &fontfamilies[findex];

**Comment by mbudnek** [comment]
Yes, recompiling with that patch eliminated the slowdown.

**Comment by mintty** [comment]
Released 2.8.2
